caddyhttp/basicauth/basicauth.go

```diff
@@ -63,7 +63,7 @@ func (a BasicAuth) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error
 
 			// let upstream middleware (e.g. fastcgi and cgi) know about authenticated
 			// user; this replaces the request with a wrapped instance
-			r = r.WithContext(context.WithValue(r.Context(),
+			*r = *r.WithContext(context.WithValue(r.Context(),
```

Comment:
@mholt  @tobya  
The problem is that within the scope of `basicauth`  the pointer to `http.Request` is being changed as `WithContext` allocates a new `r (r2)`.   Thus the dereference is needed.

if we're concerned about race issues in goroutines,  then I think we need to capture the Authorization header username before the middleware chain, to have `{user}` placeholder available

Perhaps something like this [here before r](https://github.com/mholt/caddy/blob/master/caddyhttp/httpserver/server.go#L295) is passed down to middlwareChain?   Then `{user}` is available to all downstream middleware
Quck testing looks good so far,  can look into it more later.
```diff
--- a/caddyhttp/httpserver/server.go
+++ b/caddyhttp/httpserver/server.go
@@ -292,6 +292,12 @@ func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {

        w.Header().Set("Server", "Caddy")
        c := context.WithValue(r.Context(), staticfiles.URLPathCtxKey, r.URL.Path)
+       // if r has an Authorization header add username
+       // to context for {user} placeholder in downstream middleware
+       username, _, ok := r.BasicAuth()
+       if ok {
+               c = context.WithValue(c, RemoteUserCtxKey, username)
+       }
        r = r.WithContext(c)
```